### PR TITLE
[Bugfix]Fix MiniCPM's LoRA bug

### LINKED ISSUE #9282: [Bug]: MiniCPM3-4B is support lora by --enable-lora ?
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>
docker pull vllm/vllm-openai:v0.6.2 this vllm
```text
Your output of `python collect_env.py` here
```

</details>

### Model Input Dumps

_No response_

### 🐛 Describe the bug

when i use this 
image
then get the error
image
Does this model currently not support Lora but i read the describe
image
it is supported

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by jeejeelee**
FILL IN THE PR DESCRIPTION HERE

FIX #9282  
ping @DarkLight1337  

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by DarkLight1337** [comment]
@ML-GCN can you try out this PR and see if it fixes the issue for you?

**Comment by ML-GCN** [comment]
> @ML-GCN can you try out this PR and see if it fixes the issue for you?

ok i try now

**Comment by ML-GCN** [comment]
> @ML-GCN can you try out this PR and see if it fixes the issue for you?

not yet
image
from the error shows that only like 'q_proj', 'k_proj', 'v_proj', 'o_proj', 'gate_proj', 'up_proj', 'down_proj' ... this linear layer 
but the cpm3-4b Trainable linear layer is
 "kv_a_proj_with_mqa",
    "up_proj",
    "q_b_proj",
    "kv_b_proj",
    "o_proj",
    "gate_proj",
    "down_proj",
    "q_a_proj"

**Comment by jeejeelee** [comment]
@ML-GCN  could you provibe your lora config?

**Comment by ML-GCN** [comment]
> @ML-GCN could you provibe your lora config?

{
  "alpha_pattern": {},
  "auto_mapping": null,
  "base_model_name_or_path": "/dataNfs/pre-trained/MiniCPM3-4B",
  "bias": "none",
  "fan_in_fan_out": false,
  "inference_mode": true,
  "init_lora_weights": true,
  "layer_replication": null,
  "layers_pattern": null,
  "layers_to_transform": null,
  "loftq_config": {},
  "lora_alpha": 16,
  "lora_dropout": 0.1,
  "megatron_config": null,
  "megatron_core": "megatron.core",
  "modules_to_save": null,
  "peft_type": "LORA",
  "r": 8,
  "rank_pattern": {},
  "revision": null,
  "target_modules": [
    "kv_a_proj_with_mqa",
    "up_proj",
    "q_b_proj",
    "kv_b_proj",
    "o_proj",
    "gate_proj",
    "down_proj",
    "q_a_proj"
  ],
  "task_type": "CAUSAL_LM",
  "use_dora": false,
  "use_rslora": false
}

**Comment by jeejeelee** [comment]
@ML-GCN   Please try it again

**Comment by ML-GCN** [comment]
> @ML-GCN can you try out this PR and see if it fixes the issue for you?

i try the first commit i found the output is error by origin params MiniCPM3-4B (https://huggingface.co/openbmb/MiniCPM3-4B)

example i input 你好 but the output is 
image
Can you take a look？thanks you

**Comment by ML-GCN** [comment]
> @ML-GCN Please try it again

There was no error reported  but the output is error  As I mentioned above

**Comment by jeejeelee** [comment]
It seems this PR is not responsible for the unsatisfactory result.

the main branch:
```text
Prompt: 'Hello, my name is', Generated text: " [Your Name], and I'm a student in the field of computer science"
Prompt: 'The president of the United States is', Generated text: ' the head of state and the head of government of the United States. The role'
Prompt: 'The capital of France is', Generated text: ' Paris.'
Prompt: 'The future of AI is', Generated text: ' often associated with two categories: strong AI and weak AI.\n\n- Strong'
Prompt: '你好', Generated text: '！ 区隔理论是古斯塔夫.勒庞在《乌合'
```
current PR branch:

```text
Prompt: 'Hello, my name is', Generated text: " [Your Name], and I'm a student in the field of computer science"
Prompt: 'The president of the United States is', Generated text: ' the head of state and the head of government of the United States. The role'
Prompt: 'The capital of France is', Generated text: ' Paris.'
Prompt: 'The future of AI is', Generated text: ' often associated with two categories: strong AI and weak AI.\n\n- Strong'
Prompt: '你好', Generated text: '！ 区隔理论是古斯塔夫.勒庞在《乌合'

```

**Comment by ML-GCN** [comment]
> It seems this PR is not responsible for the unsatisfactory result.
> 
> the main branch:
> 
> ```
> Prompt: 'Hello, my name is', Generated text: " [Your Name], and I'm a student in the field of computer science"
> Prompt: 'The president of the United States is', Generated text: ' the head of state and the head of government of the United States. The role'
> Prompt: 'The capital of France is', Generated text: ' Paris.'
> Prompt: 'The future of AI is', Generated text: ' often associated with two categories: strong AI and weak AI.\n\n- Strong'
> Prompt: '你好', Generated text: '！ 区隔理论是古斯塔夫.勒庞在《乌合'
> ```
> 
> current PR branch:
> 
> ```
> Prompt: 'Hello, my name is', Generated text: " [Your Name], and I'm a student in the field of computer science"
> Prompt: 'The president of the United States is', Generated text: ' the head of state and the head of government of the United States. The role'
> Prompt: 'The capital of France is', Generated text: ' Paris.'
> Prompt: 'The future of AI is', Generated text: ' often associated with two categories: strong AI and weak AI.\n\n- Strong'
> Prompt: '你好', Generated text: '！ 区隔理论是古斯塔夫.勒庞在《乌合'
> ```
ok  i try again  thanks you

**Comment by jeejeelee** [comment]
I also tested the generated result using https://huggingface.co/openbmb/MiniCPM3-4B#inference-with-vllm. The result is :
```test
你好，我是MiniCPM系列模型，由面壁智能和OpenBMB开源社区开发。详细信息请访问 https://github.com/OpenBMB/
```

**Comment by ML-GCN** [comment]
> I also tested the generated result using https://huggingface.co/openbmb/MiniCPM3-4B#inference-with-vllm. The result is :
> 
> ```
> 你好，我是MiniCPM系列模型，由面壁智能和OpenBMB开源社区开发。详细信息请访问 https://github.com/OpenBMB/
> ```

yes this is right  but you first  test the input is 你好 is error result 
image

**Comment by jeejeelee** [comment]
@ML-GCN The first test is based on https://github.com/vllm-project/vllm/blob/main/examples/offline_inference.py, possibly due to the lack of the ` chat template`.

**Comment by ML-GCN** [comment]
> @ML-GCN The first test is based on https://github.com/vllm-project/vllm/blob/main/examples/offline_inference.py, possibly due to the lack of the ` chat template`.

i found the diff <s> Can you check which part of the modified code caused it  thanks you

vllm 0.6.2:
the prompt is 
<|im_start|>user\n你好<|im_end|>\n<|im_start|>assistant\n

vllm 0.6.2 + you commit 
the prompt is  
<s><|im_start|>user\n你好<|im_end|>\n<|im_start|>assistant\n

mybe this Caused by <s>

**Comment by ML-GCN** [comment]
> @ML-GCN The first test is based on https://github.com/vllm-project/vllm/blob/main/examples/offline_inference.py, possibly due to the lack of the ` chat template`.

i found the diff prompt current PR branch will new add the tokens of <s>
like is :  prompt: '<s><|im_start|>user\n你好<|im_end|>\n<|im_start|>assistant\n'
the right prompt is '<|im_start|>user\n你好<|im_end|>\n<|im_start|>assistant\n'

**Comment by jeejeelee** [comment]
@ML-GCN sry,   I don't quite understand what you mean.

**Comment by ML-GCN** [comment]
> @ML-GCN sry, I don't quite understand what you mean.
the main branch:
prompt is 
image
current PR branch:
image
Why is there an additional token （s）at the beginning

**Comment by ML-GCN** [comment]
> @ML-GCN sry, I don't quite understand what you mean.

o  it is my error  I read it wrong  
sorry

**Comment by jeejeelee** [comment]
> > @ML-GCN sry, I don't quite understand what you mean.
> 
> o it is my error I read it wrong sorry


No worries. Does this PR resolve your previous issue?

**Comment by ML-GCN** [comment]
> > > @ML-GCN sry, I don't quite understand what you mean.
> > 
> > 
> > o it is my error I read it wrong sorry
> 
> No worries. Does this PR resolve your previous issue?

yes  thanks you very much 
When will it be updated to Docker

**Review by ML-GCN — COMMENTED** [review]
this pr sove my  request  thank you

**Comment by DarkLight1337** [comment]
The introduction of ChatGLM is causing LoRA tests to fail on main. Can you fix it in this PR as well? I think you need to update `LoRAModelManager`.

```diff,py
         # Used to indicate whether the model is a multimodal model
-        self.supports_mm: bool = supports_multimodal(self.model)
+        self.supports_mm = (supports_multimodal(self.model)
+                            # In case the model only supports LoRA for
+                            # text modules (e.g. ChatGLM)
+                            and hasattr(self.model, "get_mm_mapping"))
```

**Comment by jeejeelee** [comment]
@DarkLight1337   I have made the modifications. I don't understand why GLM4V is integrated in this way, it a bit confusing. I think it would be better to use different model script to distinguish between VL and LLM.

**Comment by DarkLight1337** [comment]
> @DarkLight1337 I have made the modifications. I don't understand why GLM4V is integrated in this way, it a bit confusing. I think it would be better to use different model script to distinguish between VL and LLM.

This is because they use the same architecture name, so we can't distinguish between them when mapping them from HF to vLLM implementation.

**Review by DarkLight1337 — APPROVED** [review]
LoRA tests pass should this should be good to merge.
